### ISSUE TITLE
Learning Rust via Advent of Code

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2678]
Learning Rust via Advent of Code<br>
<br>
http://bit.ly/2GrpXoS<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<h1>Learning Rust via Advent of Code</h1>
<p>February 3, 2019</p>
</div>
<div>
<h2>Learning Rust via Advent of Code</h2>
<p>For years the games industry has looked at Rust as the promised land. The language that will save us from C++.</p>
<p>In 2018 I decided to finally take some time to learn Rust. I read the <a href="https://doc.rust-lang.org/book/">book</a>; which was well organized and delightful. I attended <a href="https://rustconf.com/">Rust Conf</a> in nearby Portland. Then I started dabbling at home. I don't know much, but I'm making progress.</p>
<h2>Advent of Code</h2>
<p>The <a href="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/(https://adventofcode.com/">Advent of Code</a> is a programming competition organized by <a href="https://twitter.com/ericwastl">Eric Wastl</a>. It's an <a href="https://en.wikipedia.org/wiki/Advent_calendar">advent calendar</a> where every day he releases a new puzzle.</p>
<p>I decided to use the competition as means to learn Rust. I'm writing two blog posts. This post focuses primarily on my experience with Rust. My other post focuses on puzzle solutions and optimizations. If you'd like to read that post it's available here:</p>
<p><a href="https://www.forrestthewoods.com/blog/solving-advent-of-code-in-under-a-second/">Solving Advent of Code in Under a Second</a></p>
<h2>Advent Experience</h2>
<p>Learning a new language via Advent of Code turned out to be a <i>great</i> idea. I highly recommend it.</p>
<p>People use every language under the sun to solve AoC puzzles. Python, JavaScript, C, C++, C#, Rust, Haskell, Clojure, Perl, Fortran, Lisp, Lua, Scala, OCaml, Ruby, Php, Swift, and more. Many folks were, like me, using AoC to learn something new.</p>
<p>The <a href="https://www.reddit.com/r/adventofcode/">community</a> is great. Thousands of programmers are working to solve the same problems. The subreddit has a solutions thread for every puzzle. If you get stuck you can ask questions or peek at other solutions.</p>
<p>Once you have a solution it's useful to see what other people did. I learned a lot of Rust tricks and new algorithms this way.</p>
<h2>Rust Noob</h2>
<p>I'm a total noob at Rust. I've written two or three thousand lines on toy projects. This post is not "how to write good Rust code". I don't know what I'm doing! This is about my trials, tribulations, and lessons learned.</p>
<p>My solutions and full project are linked below. Solutions for all 25 days are in one file so it's easy to skim.</p>
<p><a href="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/adventofcode2018_forrestsmith_rust">Source Code</a><br>
<a href="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/AdventOfCode2018_ForrestSmith_Rust.zip">Project</a></p>
<h2>Parsing Text is Hard</h2>
<p>My first stumbling block was parsing. Almost every AoC problem starts with parsing lines of text from an input file.</p>
<p>String parsing in Rust is not easy. There's no random access into Rust Strings because they're UTF-8 encoded. The chars iterator is not very ergonomic. Using <code>as_bytes</code> helps only a little.</p>
<p>My first few attempts were complicated and inefficient. Then I learned how to use Rust's regex library by looking at <a href="https://github.com/BurntSushi/advent-of-code">BurntSushi's AoC solutions</a>.</p>
<p>I'm not gonna lie. Using regex's feels like overkill. Consider the following line of input:</p>
<pre>
#1 @ 338,764: 20x24
</pre>
<p>Here's my matching Regex code.</p>
<pre>
Regex<span>::</span>new(r<span>"#\d+ @ (?P&lt;x&gt;\d+),(?P&lt;y&gt;\d+): (?P&lt;w&gt;\d+)x(?P&lt;h&gt;\d+)"</span>).unwrap();
</pre>
<p>Is there a better expression? I don't know. Probably. I feel like there should be a helper for <i>simple</i> parse operations.</p>
<pre>
parse<span>!</span>(<span>"#{} @ {},{}: {}x{}"</span>, id, x, y, w, h);
</pre>
<p>This would be a clean inverse of <code>println!</code>. It would be great for simple cases.</p>
<h3>Rexer</h3>
<p>I'm not a regex expert. I'm not even an apprentice. I've always joked that regular expressions have the elusive write-once flag. Once you finish writing one you'll never understand what it does ever again.</p>
<p>To help write my parsing regexes I used the free website <a href="https://regexr.com/">Regexr</a>. It works well for the <a href="https://crates.io/crates/regex">Regex crate</a> which uses <a href="https://en.wikipedia.org/wiki/Perl_Compatible_Regular_Expressions">PCRE</a>-like syntax.</p>
<p>Using an interactive webtool was pleasant. It is far faster than having to compile and run Rust code.</p>
<img src="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/img/01.png" alt="Regexr Tool">
<h2>Iterators</h2>
<p>My background is C++. In C++ an "iterator" often boils down to a raw pointer. Rust iterators are <a href="https://doc.rust-lang.org/std/iter/index.html#laziness">lazy</a>, composable, and impressively powerful.</p>
<p>I spent a fair amount of time referencing <a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html">iterator API documentation</a>. I'm still learning what features are available and how to use them. I had a lot of fun trying to be clever.</p>
<h3>Enumerate</h3>
<p>I love <a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate">enumerate</a>. It provides a way to use iterators but still have an iteration count.</p>
<pre>
<span>let</span> <span>mut</span> values <span>:</span> Vec<span>&lt;</span>usize<span>&gt;</span> <span>=</span> vec<span>!</span>[<span>3</span>,<span>5</span>,<span>8</span>,<span>13</span>,<span>21</span>];
<span>for</span> (i, value) in values.iter_mut().enumerate() {
    <span>*</span>value <span>*=</span> i;
}  
assert_eq<span>!</span>(values, [<span>0</span>,<span>5</span>,<span>16</span>,<span>39</span>,<span>84</span>]);
</pre>
<h3>Custom Iterators</h3>
<p>Multiple puzzles involved grids and neighbor cells. It was common to perform some action for each neighbor. Iterating neighbors requires checking for borders and possibly neighbor value.</p>
<p>I wrote my own AdjacencyIterator to iterate neighbors and perform border checks. It was super simple. Way easier than writing a C++ <a href="https://stackoverflow.com/questions/8054273/how-to-implement-an-stl-style-iterator-and-avoid-common-pitfalls">STL-compatible iterator</a>.</p>
<p>Rust isn't quite as nice or flexible as C# <code>IEnumerable</code>. Maybe someday?</p>
<h3>Expressiveness</h3>
<p>Adjacency iterators plus built-in features enabled some delightfully expressive code. Consider the following:</p>
</div>
<div>
<pre>
<span>let</span> slot <span>=</span> units  
   .iter()
   .filter(<span>|</span>unit<span>|</span> self.is_enemy(u)                      <span>// ignore non-enemies  </span>
   .flat_map(<span>|</span>unit<span>|</span> unit.pos.adjacent_positions(w,h))   <span>// iterate adjacent positions  </span>
   .filter(<span>|</span>pos<span>|</span> <span>!</span>walls[pos] <span>&amp;&amp;</span> <span>!</span>occupied[pos]          <span>// ignore walls or occupied spaces  </span>
   .min_by_key(<span>|</span>pos<span>|</span> (self.pos <span>-</span> pos).length_sq());     <span>// pick closest slot  </span>
</pre>
</div>
<div>
<p>That code will find the closest unoccupied position that is adjacent to an enemy. If I had written that in C++ it might look like this:</p>
</div>
<div>
<pre>
<span>// track best</span>
<span>int</span> best_distance_sq <span>=</span> std<span>::</span>numeric_limits<span>&lt;</span><span>int</span><span>&gt;::</span>max();
Vec2i best_slot;
    
<span>// iterate units</span>
<span>for</span> (<span>size_t</span> i <span>=</span> <span>0</span>; i <span>&lt;</span> units.size(); <span>++</span>i) {
    Unit <span>const</span><span>&amp;</span> unit <span>=</span> units[i];
    
    <span>// ignore non-enemies</span>
    <span>if</span> (<span>this</span><span>-&gt;</span>is_enemy(unit) {
    
        <span>// iterate adjacent positions</span>
        <span>static</span> constexpr std<span>::</span>array<span>&lt;</span>Vec2i<span>&gt;</span> offsets <span>=</span> {{<span>-</span><span>1</span>,<span>0</span>}, {<span>1</span>,<span>0</span>}, {<span>0</span>,<span>-</span><span>1</span>}, {<span>0</span>,<span>1</span>}};
        <span>for</span> (<span>size_t</span> j <span>=</span> <span>0</span>; j <span>&lt;</span> offsets.size(); <span>++</span>j) {
            Vec2 pos <span>=</span> unit.pos <span>+</span> offset;
    
            <span>// check bounds</span>
            <span>if</span> (pos.x <span>&gt;=</span> <span>0</span> <span>&amp;&amp;</span> pos.x <span>&lt;</span> width <span>&amp;&amp;</span> pos.y <span>&gt;=</span> <span>0</span> <span>&amp;&amp;</span> pos.y <span>&lt;</span> height) {  
    
                <span>// ignore walls and occupied spaces</span>
                 <span>if</span> (<span>!</span>walls[pos] <span>&amp;&amp;</span> <span>!</span>occupied[pos]) {
    
                    <span>// pick closest slot</span>
                    <span>int</span> dist_sq <span>=</span> (<span>this</span><span>-&gt;</span>pos <span>-</span> pos).length_sq();
                    <span>if</span> (dist_sq <span>&lt;</span> best_distance_sq) { 
                        best_distance_sq <span>=</span> dist_sq; 
                            best_slot <span>=</span> pos;
                    }
                } 
            } 
        }
    }
} 
</pre>
</div>
<div>
<p>Those blocks of code will produce the exact same result. In theory, they should have the same performance.</p>
<p>Remember, Rust iterators are lazy. The function <code>filter</code> doesn't process every element immediately. It creates a struct that stores a closure. That closure only executes when the iterator function <code>next</code> is called. Which doesn't happen until <code>min_by_key</code>.</p>
<p>I tried to produce a <a href="https://godbolt.org/z/61a5O_">Godbolt</a> example to compare the assembly generated. Unfortunately it requires a lot of boilerplate. The output is too long to reasonably compare.</p>
<p>Writing idiomatic Rust code seems to require trusting the compiler. This is especially true for closure-based iterators. I don't love this. I've already hit instances where I needed to "unroll" simple iterator code to improve performance.</p>
<h3>Assorted Goodies</h3>
<p>I used and enjoyed quite a few iterator built-ins. Here's a brief summary of what I used for AoC.</p>
<ul>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect">collect</a> - transform iterator into a collection; such as a Vec</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count">count</a> - counts iterations to reach end</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.cycle">cycle</a> - creats iterator that loops</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate">enumerate</a> - creates iterator that returns tuple of iteration count and value</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter">filter</a> - uses closure to yield values</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find">find</a> - searches for value</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map">flat_map</a> - map plus flatten</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold">fold</a> - applies function to produce a single value</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each">for_each</a> - calls closure for each value</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map">map</a> - transforms each value into another</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max">max</a> - returns largest value in iteration</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max_by">max_by</a> - returns largest value using provided comparison function</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.min_by_key">min_by_key</a> - returns value given a function</li>
<li>
<a href="https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.sum">sum</a> - returns sum of iterator values</li>
</ul>
<h3>Compilation Woes</h3>
<p>I had to unroll some iterators just to compile. Consider the following:</p>
<pre>
dungeon.tiles  
    .iter()  
    .filter(<span>|</span>tile<span>|</span> <span>!</span>visited.contains(tile))  
    .for_each(<span>|</span>tile<span>|</span> {  
        do_thing(tile);  
        visited.insert(tile);  
    }); 
</pre>
<p>This fails to compile due to the <a href="https://doc.rust-lang.org/1.8.0/book/references-and-borrowing.html">borrow checker</a>.</p>
<p>The problem is that <code>filter</code> immutably captures visited and <code>for_each</code> mutably captures it. The Rust borrow checker does not allow this.</p>
<p>We could move <code>.contains(tile)</code> into <code>for_each</code> and it'd compile. That may even be preferable. What's interesting, and often frustrating, is that naive "unrolling" will compile while iterators do not.</p>
<h2>Profiling</h2>
<p>The Rust profiling story is mixed. It could be worse. It could be a <i>lot</i> better.</p>
<h3>Sampling</h3>
<p>I'm a big fan of <a href="https://en.wikipedia.org/wiki/Profiling_(computer_programming)#Statistical_profilers">sampling profilers</a>. Everyone should have a few in their toolkit. A good sampling profiler can give actionable information in four mouse clicks.</p>
<p>Sampling in Rust works well enough. It'd be better if the function names were demangled.</p>
<p>I tried, and was pleasantly surprised by, Visual Studio's analyzer.</p>
<img src="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/img/02.png" alt="Visual Studio Analyzer Result">
<p>VTune also worked.</p>
<img src="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/img/03.png" alt="VTune Result">
<p>I couldn't figure out how to map source files. <a href="https://imgur.com/a/V14bu">Screenshots</a> leads me to believe it's possible. I don't know how. :(</p>
<h3>Flamegraphs</h3>
<p>I have mixed opinions on <a href="http://www.brendangregg.com/flamegraphs.html">flamegraphs</a>. They're a nice visualization of sampling data. Width of each node is a proportional to total time spent. I often prefer the call tree. Your mileage may vary.</p>
<img src="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/img/04.png" alt="Canonical Flamegraph">
<p>Source: <a href="https://queue.acm.org/downloads/2016/Gregg4.svg">Brendan Gregg</a></p>
<p>The most recommended Rust profiling tool is the <a href="https://crates.io/crates/flame">flame</a> crate. The <a href="https://crates.io/crates/flamer">flamer</a> crate provides convenience macros.</p>
<p>Flame is a weird crate. It produces flamegraphs, but requires manual instrumentation. It doesn't offer a timeline or thread based view. I tried to instrument some inner-loops and it produced a 28 megabyte HTML file that wouldn't load. It's kinda the worst of both worlds.</p>
<p>I ended up using manual timers. :(</p>
<h3>Instrumention</h3>
<p>Rust has a big hole in its profiling story.</p>
<p>Flamegraphs are an important tool, but they aren't nearly enough. They aren't real time. They don't provide multithreading information. They expose a very limited type of data. They can't help with frame spikes. Webtools are janky and slow.</p>
<p>I'd love to see Rust get some professional grade profiling tools like <a href="http://www.radgametools.com/telemetry.htm">Telemetry</a> or <a href="https://www.puredevsoftware.com/framepro/index.htm">FramePro</a>.</p>
</div>
<div>
<p>C/C++ has many robust profiling options. I've used <a href="https://github.com/Celtoys/Remotery">Remotery</a> and it works well. I found several more options with a few minutes of Googling; <a href="https://www.superluminal.eu/">superluminal</a>, <a href="https://github.com/jonasmr/microprofile">microprofile</a>, <a href="https://github.com/yse/easy_profiler">easy profiler</a>, and <a href="https://github.com/bombomby/brofiler">brofiler</a>.</p>
<p><a href="https://aras-p.info/blog/2017/08/08/Unreasonable-Effectiveness-of-Profilers/">Chrome Trace</a> via <a href="https://github.com/hrydgard/minitrace">minitrace</a> is another option.</p>
</div>
<div>
<p>The lack of a good instrumented CPU profiler is a huge hole in the Rust ecosystem. It's also a huge opportunity.</p>
<h2>Benchmarking</h2>
<p>Rust's benchmarking story is mixed.</p>
<p>The goto crate is <a href="https://docs.rs/criterion/0.2.7/criterion">Criterion</a>. I had a lot of trouble with it.</p>
<p>My full solution set takes less than one-second to run. Criterion claims to be a "statistics-driven micro-benchmarking" library. Micro appears to be quite literal! Code that takes hundreds of milliseconds may be out of scope.</p>
<p>I attempted to let Criterion run overnight with default settings. It completed benchmarks for only 15 of 25 days in ~12 hours. &#129335;</p>
<p>Part of the problem is poor <a href="https://docs.rs/criterion/0.2.7/criterion/">documentation</a>. Functions are documented. How to use them is not. The user guide is insufficient.</p>
<p>I couldn't figure out how to even call <code>sample_size</code>. I had to search GitHub to find a working example. I tried to set <code>measurement_time</code>. It took more than an order of magnitude longer than specified.</p>
<p>Criterion produces some pretty charts via <a href="http://www.gnuplot.info/">gnuplot</a>. It has a lot of nice features. I was unable to effectively use the tool.</p>
<img src="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/img/07.png" alt="Criteron Graph">
<h2>Multithreading</h2>
<p>One of Rust's mottos is <a href="https://blog.rust-lang.org/2015/04/10/Fearless-Concurrency.html">fearless concurrency</a>. For AoC I chose to focus on single-threaded performance.</p>
<p>I've heard so much about <a href="https://github.com/rayon-rs/rayon">Rayon</a> I couldn't resist giving it a try. In 10 minutes I had naive multithreading working. On my 6-core/12-thread Intel i8700k it cut my total run-time from ~750ms to 200ms. Color me impressed!</p>
<p>That was naively running each puzzle in parallel. Running all solutions 100x it worked out to about 100ms on average. Not bad.</p>
<p>Proper multithreading requires a lot more than sprinkling <code>par_iter()</code> here and there. But for it to have worked so easily is a little mind blowing. I've had <i>horrible</i> experiences trying to parallelize pre-existing C++. Rayon was a breath of fresh air.</p>
<h2>Optimization</h2>
<p>I spent a fair amount of time on optimization. Some of it algorithmic and some Rust specific. For further details please see my other post:</p>
<p><a href="https://www.forrestthewoods.com/blog/solving-advent-of-code-in-under-a-second/">Solving Advent of Code in Under a Second</a></p>
<h2>Assorted Thoughts</h2>
<h3>Rust Playground</h3>
<p>The <a href="https://play.rust-lang.org/?version=stable&amp;mode=debug&amp;edition=2015&amp;gist=8cf114170682b9c1381473fca58c7952">Rust Playground</a> is great. I used it to quickly test questions. For example, what is the std::mem::size_of result for various structs?</p>
<p>It's also great for creating snippets to share in the <a href="https://www.reddit.com/r/rust">r/rust</a> newbie thread. Which I did on more than a few occassions.</p>
</div>
<div><img src="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/img/09.png" alt="Rust Playground"></div>
<div>
<h3>Debugging</h3>
<p>Rust's debugging story is "needs improvement".</p>
<p>Visual Studio Code is pretty easy to configure as a debugger. Step debugging actually works. So does the watch window.</p>
<p>Unfortunately most containers aren't inspectable. Vec works because it's basically an array. HashMap and HashSet are inscrutable.</p>
<img src="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/img/08.png" alt="VSCode Debugger">
<p>Visual Studio has <a href="https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2017">natvis</a> for C++. It's not great. I have a lot of complaints. It's way better than Rust's nothing.</p>
<p>I'm hoping Rust 2019 lays the groundwork for improved IDE and debugging in 2020.</p>
<h3>Bit Packing</h3>
<p><code>Option&lt;bool&gt;</code> is only one byte. <code>Option</code> is also one byte for an <code>enum</code> when the <code>enum</code> value doesn't need all the bits. This was a pleasant surprise.</p>
<h3>Implicit Casts</h3>
<p>Rust doesn't have implicit casts. If you want to pass a <code>u32</code> to a function that needs a <code>usize</code> you have to explicitly cast it.</p>
<p>I think I like this. It leads to annoyingly verbose code. But it did prevent at least a couple of bugs.</p>
<h3>Inference</h3>
<p>Rust is statically typed. However it has robust support for type inference.</p>
<pre>
<span>let</span> elem <span>=</span> <span>5</span><span>u8</span>;  
<span>let</span> <span>mut</span> vec <span>=</span> Vec<span>::</span>new();  
vec.push(elem);  
vec.push(<span>6</span><span>u32</span>); <span>// error: expected u8, found u32  </span>
</pre>
<p>The type of <code>vec</code> is inferred to be <code>Vec&lt;u8&gt;</code>. You can leave off a <i>shocking</i> amount of type information. For better or worse.</p>
<h3>rustfmt</h3>
<p>I enabled rustfmt for auto-formatting. I like it 95% of the time. For AoC I'm jealous of the brevity of Python. Rust is quite verbose, and rustfmt makes it a little worse.</p>
<p>Rustfmt badly needs a <code>clang-format off</code> feature for block level control.</p>
<h3>Helper Lambdas</h3>
<p>I like lambdas in C++11. I use them regularly for small helpers that exist solely within a function.</p>
<p>This is hard to do in Rust. It causes borrow checker errors similar to iterator closure issue described above.</p>
<p>What make it frustrating is that an uncompilable helper lambda will often work <i>perfectly</i> if changed to a macro. That feels like a bug to me. This case isn't quite mentioned in a recent post on <a href="http://smallcultfollowing.com/babysteps/blog/2018/11/01/after-nll-interprocedural-conflicts/">non-lexical lifetimes</a>. I think it should be.</p>
<h3>HashMap Initialization</h3>
<p>I had a hard time initializing <code>HashMap</code>. Here's what I'd like to do:</p>
<pre>
<span>let</span> allowable_gear <span>:</span> HashMap<span>&lt;</span>(TileType,TileType), <span>&amp;</span>[Gear]<span>&gt;</span> <span>=</span> {  
    (Rocky, Rocky) <span>=</span> <span>&amp;</span>[Climbing, Torch],  
    (Rocky, Wet) <span>=</span> <span>&amp;</span>[Climbing], 
    (Wet, Narrow) <span>=</span> <span>&amp;</span>[Neither],
};
</pre>
<p>This is a <code>HashMap</code> that maps a pair of <code>TileTypes</code> to an array of <code>Gear</code>. This relationship is constant and known at compile time. I couldn't find a clean, idiomatic way to do this. :(</p>
<h3>BinaryHeap</h3>
<p>The standard library provides a <a href="https://en.wikipedia.org/wiki/Binary_heap">max-heap</a>. I regularly needed a min-heap. I got one by making a custom struct with custom <a href="https://doc.rust-lang.org/beta/std/cmp/index.html">compare</a> function.</p>
<p>This was a bit of a pain. It could have been easier. Priority queues with customizable compare functions seems like a useful thing that would exist. I'm probably doing it wrong?</p>
<h3>Rc&lt;RefCell&lt;Foo&gt;&gt;</h3>
<p>I've barely touched <code>Rc</code> and <code>RefCell</code>. For day23 I wrote an Octree and it used an <code>Rc&lt;RefCell&lt;OctreeNode&gt;&gt;</code>. I wanted to typedef that to <code>OctreeNodePtr</code>. Except that didn't work for creating new values.</p>
<p>Graphs are notoriously difficult in Rust. I'm pretty sure I made my Octree incorrectly. But I'm not sure what I should have done instead. Suggestions welcome!</p>
<h3>Sorting</h3>
<p>For several puzzles I sorted a struct based on sequential fields. For example:</p>
<pre>
dates.sort_by(<span>|</span>a,b<span>|</span> { 
    <span>if</span> a.year <span>!=</span> b.year { a.year.cmp(<span>&amp;</span>b.year) }  
    <span>else</span> <span>if</span> a.month <span>!=</span> b.month { a.month.cmp(<span>&amp;</span>b.month) }  
    <span>else</span> <span>if</span> a.day <span>!=</span> b.day { a.day.cmp(<span>&amp;</span>b.day) } 
    <span>else</span> <span>if</span> a.hour <span>!=</span> b.hour { a.hour.cmp(<span>&amp;</span>b.hour) }
    <span>else</span> { a.minute.cmp(<span>&amp;</span>b.minute) }
});
</pre>
<p>I wonder if there is a nice macro crate to help with this? I'd love to write:</p>
<pre>
compare<span>!</span>(a, b, year, month, day, hour, minute);
</pre>
<p>My sort for day23 was more complex. It sorted by <code>foo.max_possible</code> then <code>foo.bounds.volume()</code> then <code>foo.bounds.distance_from(origin)</code>. Here's one idea for a useful sort API.</p>
<pre>
foos.sort_by(<span>|</span>v<span>|</span> v.max_possible)
    .then_by(<span>|</span>v<span>|</span> v.bounds())
    .then_by(<span>|</span>v<span>|</span> v.bounds.distance_from(origin))
    .sort(); 
</pre>
<p>That would be great. That API design feels pretty close to right. What do you think?</p>
<h3>Boilerplate</h3>
<p>I chose to write a lot of boilerplate in my solutions. I wanted to keep all my code in one place for each puzzle. I didn't want to use ambiguous tuples of unnamed types.</p>
<p>I wrote <code>struct Pos</code> for 8 out of 25 days. In <code>i8</code>, <code>i16</code>, and <code>i32</code> flavors. I probably should have used <a href="https://docs.rs/cgmath/0.16.1/cgmath/">cgmath</a> or <a href="https://www.nalgebra.org/rustdoc/nalgebra/index.html">nalgebra</a>. One my side projects is a Rust crate for video game 3d math. Maybe it'll be done by Advent of Code 2019.</p>
<h3>Build Profiles</h3>
<p>Cargo doesn't have custom build profiles. This blows my mind.</p>
<p>The included profiles are dev, release, bench, test, and doc. Only dev and release are used for running a program. Historically I've run with debug, internal, release, and retail.</p>
<p>There is a <a href="http://aturon.github.io/2018/04/05/workflows/">blog post</a> and <a href="https://github.com/rust-lang/rust/issues/48683">tracking issue</a> that lead me to believe custom profiles this will happen someday. It may be awhile.</p>
<h3>Link-Time Optimization</h3>
<p>Release builds set link-time optimization to false by default. This is the correct choice due to build times.</p>
<p>Enabling LTO and changing codegen units reduced my run-time from 700ms to 600ms. That's a 14% improvement from a couple of compiler flags! More details available in <a href="https://www.forrestthewoods.com/blog/solving-advent-of-code-in-under-a-second/">other post</a>.</p>
<p>When Rust ships custom build profiles I'd love to see a MaxSpeed config. <a href="https://github.com/BurntSushi/ripgrep/issues/413">RipGrep</a> doesn't ship binaries with LTO. I think the lack of custom profiles is causing a <i>lot</i> of shipped Rust code to leave free performance on the table.</p>
<h3>New Primitive Types</h3>
<p>I wish Rust provided <code>u24</code> and <code>i24</code> types. They're useful in a few scenarios. They already exist inside various audio crates. I couldn't find a robut standalone implementation.</p>
<p>I wish there was an <code>f16</code> half-precision float. There's a good <a href="https://internals.rust-lang.org/t/adding-minifloats-f24-f16-f8-as-native-types/2367">internals thread</a> on minifloats that makes me think <code>f16</code> will happen eventually.</p>
<h2>Crates</h2>
<p>I used a handful of crates. Here's a brief list.</p>
<p>I have not tried the parsing libraries <a href="https://crates.io/crates/nom">nom</a> or <a href="https://crates.io/crates/pest">pest</a>. I wonder if they would have made parsing day 24 easy? I'd like to investigate them in the future.</p>
<h3>Crate Discovery</h3>
<p>Crate discovery is non-existent. GitHub stars is not a useful metric. Downloads on crates.io is ok but not great. I looked for some replacements for <code>std::collections</code> containers. It was hard to tell which crates were worth trying.</p>
<p>I got a huge speed-up when I replaced the built-in <code>HashMap</code> with an <code>FnvHashMap</code>. Through dumb luck I found <code>Hashbrown</code>, which is even faster.</p>
<p>There needs to be a better way to find crates. This is a hard problem.</p>
<h2>Clippy</h2>
<p><a href="https://github.com/rust-lang/rust-clippy">Clippy</a> is an easy to use linter to catch common issues.</p>
<p>I hit and fixed ~25 lints. They were minor issues or recommended improvements. However I did have to disable a few lints.</p>
<h3>needless_range_loop</h3>
<p>Bad lint. I iterated <code>0..3</code> to grab components out of multiple <code>Vec2</code> structs. My struct lacked iterators. I could have implemented an iterator and used <code>zip</code>. A simple for-loop is better.</p>
<h3>too_many_arguments</h3>
<p>Generally true. I feel my use case was appropriate.</p>
<h3>map_entry</h3>
<p>Bad lint. It was straight wrong.</p>
<p>For day22 I had a HashMap of tiles. Each tile had a cost that was dependent on tiles up and to the left. I wrote this recursively.</p>
</div>
<div>
<pre>
<span>fn</span> get_tile(<span>&amp;</span><span>mut</span> self, pos<span>:</span> pos) <span>-&gt;</span> <span>&amp;</span><span>mut</span> Tile {  
    <span>if</span> <span>!</span>self.tiles.contains_key(<span>&amp;</span>pos) {  
        <span>let</span> new_value <span>=</span> compute_value(self.get_tile(left), self.get_tile(up));
        self.tiles.insert(pos, new_value);
    } 

    self.tiles.get_mut(<span>&amp;</span>pos)
}  
</pre>
</div>
<div>
<p>The lint recommended version is:</p>
</div>
<div>
<pre>
<span>fn</span> get_tile(<span>&amp;</span><span>mut</span> self, pos<span>:</span> pos) <span>-&gt;</span> <span>&amp;</span><span>mut</span> Tile {  
    self.tiles
        .entry(<span>&amp;</span>pos)
        .or_insert_with(<span>||</span> {
            <span>let</span> new_value <span>=</span> compute_value(self.get_tile(left), self.get_tile(up));
            self.tiles.insert(pos, new_value);
        });
}  
</pre>
</div>
<div>
<p>This doesn't compile. <code>entry</code> stores a reference to <code>tiles</code>. Then <code>or_insert_with</code> creates a closure with a mutable reference to <code>tiles</code> for recursion. That's not allowed.</p>
<p>My recursive pattern isn't common. I'm not surprised Clippy can't handle it. It's a shame because I really like the <code>entry</code> API with <code>or_insert</code> and <code>or_insert_with</code>.</p>
<h3>Lint Control</h3>
<p>Clippy needs a way to disable lints for individual functions, blocks, or lines of code. I don't want to turn these lints off everywhere. This is the same issue as wanting <code>clang format off</code> for rustfmt.</p>
<h2>Sentinels</h2>
<p>One feature I wish Rust had was a helper struct for sentinel values. Consider an 8-bit value where you want to special case <code>255</code> to mean "no value". Or an <code>i32</code> where <code>-1</code> means "no value". This is pretty common.</p>
<p>One Rusty, ahem, option would be an <a href="https://doc.rust-lang.org/std/option">Option</a>; <code>Option&lt;u8&gt;</code>. Another might be an <a href="https://doc.rust-lang.org/1.30.0/book/second-edition/ch06-01-defining-an-enum.html">enum</a>; <code>enum EFoo { None, Foo(u8) }</code>.</p>
<p>Both these approaches use two bytes of storage. One byte for the <code>u8</code>, and another for a flag. An <code>Option&lt;u32&gt;</code> bloats to a whopping 8 bytes. An <code>Option&lt;u64&gt;</code> is, wait for it, 16 bytes.</p>
<p>What I want is a struct that is stored as the underlying type and has a getter which returns an <code>Option&lt;T&gt;</code>. The getter returns <code>None</code> if the inner value is a specific value.</p>
<p>This feature may not be possible today. It might be when when <a href="https://github.com/rust-lang/rust/issues/44580">const generics</a> land.</p>
<p>There is <a href="https://doc.rust-lang.org/beta/std/num/struct.NonZeroU32.html">NonZeroU32</a>. It's similar-ish, but only works for zero. There isn't a <code>Non255U8</code> or, thankfully, <code>Non4294967296U32</code>.</p>
<h2>Final Thoughts</h2>
<p>I used Advent of Code as a means to learn Rust. This was a great idea and one I strongly recommend. AoC is a great way to learn any language.</p>
<p>I won't claim to know Rust. But I do know a lot more today than I did before. Progress is clear success in my book.</p>
<p>Thanks for reading.</p>
<h2>Links</h2>
<p><a href="https://www.forrestthewoods.com/blog/solving-advent-of-code-in-under-a-second/">Solving Advent of Code in Under a Second</a><br>
<a href="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/adventofcode2018_forrestsmith_rust">Source Code</a><br>
<a href="https://www.forrestthewoods.com/blog/learning-rust-via-advent-of-code/assets/AdventOfCode2018_ForrestSmith_Rust.zip">Project</a></p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via www.forrestthewoods.com http://bit.ly/2UD9L83<br>
<br>
February 5, 2019 at 07:40AM
